---
- name: Windows Check snap and patch and revert if failure
  hosts: all
  tasks:
  - name: Check for missing updates
    win_updates: 
      category_names: "{{ patch_cat }}"
      state: searched
    register: update_count
    
  - when: update_count.found_update_count >= 1
    block:
    - name: create snapshot
      vmware_guest_snapshot:
        datacenter: "{{ datacenter_name }}"
        folder: "{{ folder }}"
        name: "{{ inventory_hostname }}"
        state: present
        snapshot_name: pre-upgrade
      connection: local
      become: false
      
    - name: Apply missing updates
      win_updates:
        category_names: "{{ patch_cat }}"
        state: installed
      register: update_result 

    - name: 1st windows reboot
      win_reboot:
      when: update_result.reboot_required

    - name: Check for missing updates round 2
      win_updates: 
        category_names: "{{ patch_cat }}"
        state: searched
      register: update_count_2

    - when: update_count_2.found_update_count >= 1
      block:
      - name: Apply missing updates round 2
        win_updates:
          category_names: "{{ patch_cat }}"
          state: installed
        register: update_2_result

      - name: 2nd windows reboot
        win_reboot:
        when: update_2_result.reboot_required

      - name: Check for missing updates round 3
        win_updates: 
          category_names: "{{ patch_cat }}"
          state: searched
        register: update_count_3
        
      - when: update_count_3.found_update_count >= 1
        block:
        - debug:
            msg: still missing updates please investigate
        
    rescue:
     - name: restore snapshot
       vmware_guest_snapshot:
         datacenter: "{{ datacenter_name }}"
         folder: "{{ folder }}"
         name: "{{ inventory_hostname }}"
         state: revert
         snapshot_name: pre-upgrade
       connection: local
       become: false
    
     - name: power on guest
       vmware_guest:
         name: "{{ inventory_hostname }}"
         state: poweredon
       connection: local
       become: false
    
     - debug:
         msg: "{{ inventory_hostname }} failed and was restored from its snapshot"
